Panic during settlement

Possible way to reproduce: on https://github.com/vegaprotocol/vega-market-sim/ checkout `origin/undistributed_liquidity_fee_after_settle` and run `LiquidityFeesCheck.ipynb` . 

The notebook tells you where to look for log files, crash happens in cell 23 - obvious in cell 24 when python doesn't get the answers it expects. 

The core error log is:
```
(sim) D-MBP-M1:vega-market-sim davidsiska$ cat /var/folders/67/mjxp58z56yj83_1x0372gkwr0000gn/T/tmp7lne2k0m/node.err 
panic: unknown subscriber ID 2

goroutine 2384 [running]:
code.vegaprotocol.io/vega/core/oracles.(*specSubscriptions).removeSubscriber(0x140003fc4c0, 0x2)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/oracles/engine_subscribers.go:160 +0x454
code.vegaprotocol.io/vega/core/oracles.(*Engine).Unsubscribe(0x140011b5050, {0x106b2dc60, 0x1400d1fccc0}, 0x19296d4c00000003?)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/oracles/engine.go:143 +0x3c
code.vegaprotocol.io/vega/core/oracles.(*Engine).Subscribe.func1({0x106b2dc60?, 0x1400d1fccc0?}, 0x0?)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/oracles/engine.go:136 +0x38
code.vegaprotocol.io/vega/core/products.(*Future).UnsubscribeTradingTerminated(0x1400c26fd58?, {0x106b2dc60?, 0x1400d1fccc0?})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/products/future.go:48 +0x3c
code.vegaprotocol.io/vega/core/markets.(*Instrument).UnsubscribeTradingTerminated(...)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/markets/instrument.go:98
code.vegaprotocol.io/vega/core/markets.(*Instrument).Unsubscribe(0x1400d13f500, {0x106b2dc60, 0x1400d1fccc0})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/markets/instrument.go:106 +0x3c
code.vegaprotocol.io/vega/core/execution.(*Market).closeCancelledMarket(0x1400d17c000, {0x106b2dc60, 0x1400d1fccc0}, {0x1400da82b80?, 0x40?, 0x1?})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/execution/market.go:839 +0x3c
code.vegaprotocol.io/vega/core/execution.(*Market).tradingTerminated(0x1400d17c000, {0x106b2dc60, 0x1400d1fccc0}, 0x68?)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/execution/market.go:3229 +0x51c
code.vegaprotocol.io/vega/core/products.(*Future).setTradingTerminated(0x1400db323a0?, {0x106b2dc60?, 0x1400d1fccc0?}, 0xf0?, {0x0?, 0x0?})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/products/future.go:191 +0x158
code.vegaprotocol.io/vega/core/products.(*Future).updateTradingTerminated(0x1400d145710, {0x106b2dc60, 0x1400d1fccc0}, {{0x1400db323a0?, 0x1400d14dc20?, 0x1400d14dc50?}, 0x1400d1fccf0?})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/products/future.go:161 +0xc4
code.vegaprotocol.io/vega/core/oracles.(*Engine).BroadcastData(0x140011b5050, {0x106b2dc60, 0x1400d1fccc0}, {{0x1400db323a0?, 0x20?, 0x40?}, 0x1400d1fccf0?})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/oracles/engine.go:114 +0x5a8
code.vegaprotocol.io/vega/core/processor.(*App).DeliverSubmitOracleData(0x14001080dc0, {0x106b2dc60, 0x1400d1fccc0}, {0x106b3f2d0, 0x1400db36180})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/processor/abci.go:1301 +0x1c0
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).DeliverTx(0x14000ff5e40, {{0x1400d0bef00, 0x9d, 0x9d}})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:157 +0x600
code.vegaprotocol.io/vega/cmd/vega/node.(*appW).DeliverTx(0x140001b4d80?, {{0x1400d0bef00?, 0x28?, 0x1400c2713e8?}})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/cmd/vega/node/app_wrapper.go:41 +0x4c
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0x14000f86480)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:137 +0xd0
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleTransaction(0x14000f86480, {0x1400d0bef00, 0x9d, 0x9d})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:158 +0x21c
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).SendTransactionSync(0x106b2dc60?, {0x1400da41590?, 0x6?}, {0x1400d0bef00?, 0x1400be755f8?, 0x10480c68c?})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:342 +0x2c
code.vegaprotocol.io/vega/core/blockchain.(*Client).SubmitTransactionSync(0x140010067b0, {0x106b2dc60, 0x1400da41590}, 0x30?)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/blockchain/client.go:137 +0x110
code.vegaprotocol.io/vega/core/api.(*coreService).SubmitTransaction(0x140002cc140, {0x106b2dc60, 0x1400da41590}, 0x1400da15dc0)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/api/core.go:150 +0x9c
code.vegaprotocol.io/vega/protos/vega/api/v1._CoreService_SubmitTransaction_Handler.func1({0x106b2dc60, 0x1400da41590}, {0x106925780?, 0x1400da15dc0})
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/protos/vega/api/v1/core_grpc.pb.go:270 +0x74
code.vegaprotocol.io/vega/core/api.remoteAddrInterceptor.func1({0x106b2dc60, 0x1400da41530}, {0x106925780, 0x1400da15dc0}, 0x1400da356e0, 0x1400da0dce0)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/api/grpc.go:210 +0x4a8
code.vegaprotocol.io/vega/protos/vega/api/v1._CoreService_SubmitTransaction_Handler({0x106a63e40?, 0x140002cc140}, {0x106b2dc60, 0x1400da41530}, 0x1400d908d90, 0x1400002c8f0)
	/Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/protos/vega/api/v1/core_grpc.pb.go:272 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0x140002ce000, {0x106b3b300, 0x140000304e0}, 0x1400d6d3320, 0x140002426c0, 0x107cd3b20, 0x0)
	/Users/davidsiska/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1295 +0x9c4
google.golang.org/grpc.(*Server).handleStream(0x140002ce000, {0x106b3b300, 0x140000304e0}, 0x1400d6d3320, 0x0)
	/Users/davidsiska/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1636 +0x82c
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/Users/davidsiska/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:932 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/davidsiska/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:930 +0x290
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -42,6 +42,7 @@
 - [6008](https://github.com/vegaprotocol/vega/issues/6008) - Fix equity like share and average trade value calculation with opening auctions
 - [6040](https://github.com/vegaprotocol/vega/issues/6040) - Fix protocol upgrade transaction submission and small Visor improvements
 - [5977](https://github.com/vegaprotocol/vega/issues/5977) - Fix missing block height and block time on stake linking API
+- [6054](https://github.com/vegaprotocol/vega/issues/6054) - Fix panic on settlement
 
 ## 0.54.0
 

--- a/core/execution/market.go
+++ b/core/execution/market.go
@@ -839,7 +839,8 @@ func (m *Market) cleanMarketWithState(ctx context.Context, mktState types.Market
 }
 
 func (m *Market) closeCancelledMarket(ctx context.Context, t time.Time) error {
-	m.tradableInstrument.Instrument.Unsubscribe(ctx)
+	// we got here because trading was terminated so we've already unsubscribed that oracle data source.
+	m.tradableInstrument.Instrument.UnsubscribeSettlementPrice(ctx)
 
 	err := m.cleanMarketWithState(ctx, types.MarketStateCancelled)
 	if err != nil {

